components/tidb_query/src/codec/mysql/json/json_merge.rs

```diff
@@ -1,51 +1,78 @@
 // Copyright 2017 TiKV Project Authors. Licensed under Apache-2.0.
 
-use super::Json;
+use super::{Json, JsonType};
+use crate::codec::{Error, Result};
+use std::collections::BTreeMap;
 
 impl Json {
-    // `merge` is the implementation for JSON_MERGE in mysql
-    // https://dev.mysql.com/doc/refman/5.7/en/json-modification-functions.html#function_json-merge
-    //
-    // The merge rules are listed as following:
-    // 1. adjacent arrays are merged to a single array;
-    // 2. adjacent object are merged to a single object;
-    // 3. a scalar value is autowrapped as an array before merge;
-    // 4. an adjacent array and object are merged by autowrapping the object as an array.
-    pub fn merge(self, to_merge: Json) -> Json {
-        match (self, to_merge) {
-            // rule 1
-            (Json::Array(mut array), Json::Array(mut sub_array)) => {
-                array.append(&mut sub_array);
-                Json::Array(array)
-            }
-            // rule 3, 4
-            (Json::Array(mut array), suffix) => {
-                array.push(suffix);
-                Json::Array(array)
-            }
-            // rule 2
-            (Json::Object(mut obj), Json::Object(sub_obj)) => {
-                for (sub_key, sub_value) in sub_obj {
-                    let v = if let Some(value) = obj.remove(&sub_key) {
-                        value.merge(sub_value)
-                    } else {
-                        sub_value
-                    };
-                    obj.insert(sub_key, v);
+    /// `merge` is the implementation for JSON_MERGE in mysql
```

Comment:
implement for JsonRef?

Reply:
From the perspective of callers, I think it's ok just implement `merge` for `Json`. And we might be unable to gain many benefits from implementing `merge` for `JsonRef`.